### ISSUE TITLE
RTK Query: Subscription leak in React using useQuery() with skip parameter

### ISSUE THREAD

**Issue by mjwvb** [reduxjs/redux-toolkit#3182]
There seems to be a subscription leak in RTK Query (1.9.2) with React (18.2). 

When using useQuery / useQuerySubscription with the skip parameter, some of the subscriptions randomly won't go away when the component/hook unmounts or the args change. This leak is very common in our application and causes issues with some of our code in the onCacheEntryAdded endpoint lifecycle method, as its lifecycle is never ended for the leaked subscriptions.

I managed to reproduce the leak in the pokemon example sandbox by brute forcing different subscription mounts on the same endpoint (with keepUnusedDataFor=1) while toggling the skip parameter on/off at a 10ms interval. By pressing the mount button in the sandbox you can initiate the brute force action and watch the active list of subscriptions. You can see that some of the subscriptions will never go away and have become ghost subscriptions. (Of course I turned off the api calls to not ddos the pokemon database.)

Obviously the sandbox is not a real world use case, but it shows there's something wrong. In our own application it happens by simply having many different endpoints in a complex/busy react environment. However it's worth noting that we can somehow fix it by turning off the autoBatchEnhancer, unsure why...

[![Edit dry-river-ltbu89](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/dry-river-ltbu89?fontsize=14&hidenavigation=1&theme=dark)

![subscription leak](https://user-images.githubusercontent.com/9107835/219405158-ade8cdfa-c8c9-4ecd-8327-93a85ea86cec.gif)

**Comment by markerikson** [comment]
Interesting! Don't immediately have an idea why this would be happening, but thanks for the report and the repro!

**Comment by mjwvb** [comment]
Thanks for the quick reply! I'm actually a little bit unsure whether this is the actual cause of the subscription leakage in our specific situation, as it is fixed by disabling the autoBatchEnhancer. However the conditions and the result is exactly the same so it is related at least.

**Comment by markerikson** [comment]
Huh. At first glance, there's _very_ different behavior between Chrome and Firefox here. With Chrome, clicking "Unmount" makes the subscription count revert back to about 1.  With Firefox, it only goes down a couple and leaves most of them in place.

**Comment by markerikson** [comment]
@mjwvb : hmm. I actually see this happening in the sandbox, _without_ having the `autoBatchEnhancer` turned on...

**Comment by mjwvb** [comment]
In practice the leakage occurs only very randomly and with multiple useQuery() simultaneously when the event loop / cpu is too busy, so very hard to reproduce. I suspect the autoBatchEnhancer does something with the processing order of things, so the chances of a leak happening is much higher when it's turned on.

I noticed very different behavior as well, e.g. happens more often on windows chrome than on macos chrome. Maybe it has to do with the interval timing and how efficient everything is processing.

**Comment by markerikson** [comment]
Been trying to dig through these two replays.  I did use a local RTK test build from the #3188 branch.  That actually _fixes_ this issue on Chrome, but I still see the problem in FF.

Here's the recordings:

- [RTK 3182 replay: subscriptions not removing (Firefox)](https://app.replay.io/recording/rtk-3182-subscriptions-not-removing-firefox-skip-fix--d7fd7519-d8ad-4f79-81fa-14c10d3c7730?point=5516815412374688459571444351763321&focusRegion=eyJiZWdpbiI6eyJwb2ludCI6IjQ1NDMyNTk3NTEyNDc2NTI2ODAzMzU3NzM3NDI5MjQzMTkiLCJ0aW1lIjoxNjI4fSwiZW5kIjp7InBvaW50IjoiNTg0MTMzMzk2NjA5NTE5MzA5MTg0OTMzNDM0MDcxNTIyMiIsInRpbWUiOjE5ODV9fQ%253D%253D&time=1892)
- [RTK 3182 replay: subscriptions not removing (Chrome)](https://app.replay.io/recording/rtk-3182-subscriptions-not-removing-chrome-skip-fix--9dac50ca-fb6f-40d5-817c-6e5f55ed2ea5?point=3245185536631963630476737443266674&time=1644&focusRegion=eyJiZWdpbiI6eyJwb2ludCI6IjMyNDUxODU1MzY2MDgzODA2MjExMDA0MTI0MTAyNjU3ODgiLCJ0aW1lIjoxNjM0fSwiZW5kIjp7InBvaW50IjoiNDIxODc0MTE5Nzk0MTIxMDU4MzA2NjA3ODg5MDQyNjg5NiIsInRpbWUiOjE4NTh9fQ%253D%253D)

**Comment by markerikson** [comment]
@mjwvb : good news! I _think_ I've got a fix, and the actual reason why this is happening.

The `skipToken` fix in #3188 is part of it, but the other buggy bit is this line here:


https://github.com/reduxjs/redux-toolkit/blob/43d94a0b6de88c85c3e7c50bb53f2f4c58e88f4f/packages/toolkit/src/query/react/buildHooks.ts#L744-L746

Notice that it's always clearing out this promise ref any time `subscriptionRemoved` changes.

The _cause_ sequence appears to be:

- Start with an existing subscription
- Change the query arg with 
- Query hook runs with `subscriptionRemoved: true`, clears promise ref, initiates subscription (good)
- Component re-renders (status, etc)
- Query hook runs with `subscriptionRemoved: false`, clears promise ref, initiates subscription (bad)

The problem is that we _shouldn't_ be clearing out the promise ref if the subscription hasn't been removed.

Adding `if (subscriptionRemoved)` inside that effect fixes things.

Currently working on a test that helps prove the fix works.

**Comment by markerikson** [comment]
@mjwvb : could you try out the CodeSandbox preview build from #3188 and let me know if that fixes the issue on your side?

**Comment by mjwvb** [comment]
Nice! Tested the sandbox in multiple browsers and it really seems to be fixed right now. I also installed the specific commit in our application where the issue originally occurred, and now I can keep the autoBatchEnhancer turned on and have no issues at all.

Thanks for fixing the issue so quickly! Such leaks are always a struggle to pinpoint. Having a tool like replay.io seems like a blessing for those kind of issues.

**Comment by harry-gocity** [comment]
Just wanted to say thanks both @mjwvb for the repro on this and @markerikson for the fix - we've had this issue come up in a very similar and subtle scenario that I've been scratching my head over and struggling to create a repro for.

In our case, we have a set of queries in our app that use args selected from the store, and a mutation implemented with a `queryFn`. In the mutation `queryFn`, prior to returning success data we change the same store values that are used by the queries. When we trigger the mutation in the app, we unwrap the result and if successful immediately route to a different page, where the other queries are not used. But somehow, in between the dispatch inside `queryFn` and the actual router push which unmounts everything, the queries were all refetching with the updated values from the store, and the subscriptions were persisting across the route change, causing problems on the next page. 

Bumping to 1.9.5 fixes this for us 🙌🏻

**Comment by gtrak** [comment]
I'm seeing something like this on 1.9.5, react 18.2, independent of the skip param. 

What's happening is when using polling, often polling is not updated on an args change. I did a bit of digging and found that clearing behavior is related to the number of subscriptions in #1933 , and I am seeing more than one subscription in redux state with a single active useQuery hook on the page.  I confirmed all the params and args state was correct. I was able to work around it by doing my own polling with a custom hook.

Could there be more to fix around subscriptions?
